### ISSUE TITLE
R API - Last layer's output

### ISSUE THREAD

**Issue by jacquespeeters** [dmlc/mxnet#2535]
Hi, 

I'm mainly an R user and a huge fan of XGBOOST, therefore it was natural to me to give a try to MxNet. However i dod struggle a lot to accomplish anything good. So few people use MxNet-R that i feel a bit lost.

https://github.com/dmlc/mxnet/blob/master/R-package/vignettes/classifyRealImageWithPretrainedModel.Rmd I've performed this tutorial. I would like to get the second last layer's features output of the pre-trained model available. Then i'll apply the classification task given my personal problem (not fine even fine-tuning, just the outputs). I've read the manual several times but i haven't found anything and i'm not even sure this is even possible. 

Even if it's possible, i know i won't get the best results but i think it will still be better than what i got right now. Here is the student competition i participate in. 
https://inclass.kaggle.com/c/data-science-game-2016-online-selection/

All advices are welcome!

**Comment by l4mun** [comment]
Hey,

https://github.com/dmlc/mxnet/blob/master/example/notebooks/predict-with-pretrained-model.ipynb
I've found exactly what i wanted in Python, but some commands seems to be missing :(

Any help still welcomed!

**Comment by thirdwing** [comment]
I think the function you are looking for is `mxnet:::mx.model.train()`.

https://github.com/dmlc/mxnet/blob/master/R-package/R/model.R#L94

This is not exported.

**Comment by l4mun** [comment]
Thank you for your kind advice. Tried but as soon i solve an error, another one appear. I guess this is not as simple as with Python yet :/

kvstore <- mxnet:::mx.model.create.kvstore("local", model$arg.params, length(ctx), verbose=verbose)
optimizer <- mx.opt.create("sgd", rescale.grad=(1/batchsize), learning.rate=0.1)
X=rbind(normed, normed)
if (is.array(X) || is.matrix(X)) {
  if (array.layout == "auto") {
    array.layout <- mx.model.select.layout.train(X, y)
  }
  if (array.layout == "rowmajor") {
    X <- t(X)
  }
}
X <- mxnet:::mx.model.init.iter(X, 1:2, batch.size=1, is.train=TRUE)
if (!X$iter.next()) {
  X$reset()
  if (!X$iter.next()) stop("Empty input")
}

feature_extractor = mxnet:::mx.model.train(ctx = list(mx.cpu()), symbol=model$symbol, arg.params=model$arg.params, 
                                           aux.params=model$aux.params, input.shape = dim((X$value())$data), 
                                           kvstore=kvstore, optimizer = optimizer, begin.round=1, end.round=2,
                                           metric = mx.metric.accuracy, train.data=X)
Start training with 1 devices
[22:12:51] D:\chhong\mxnet\dmlc-core\include\dmlc/logging.h:235: [22:12:51] d:\chhong\mxnet\src\operator\./convolution-inl.h:322: Check failed: (dshape.ndim()) == (4) Input data should be 4D in batch-num_filter-y-x
Error: InferShape Error in conv_1: [22:12:51] d:\chhong\mxnet\src\operator\./convolution-inl.h:322: Check failed: (dshape.ndim()) == (4) Input data should be 4D in batch-num_filter-y-x

**Comment by pyd1data** [comment]
I have exactly the same problem, while  in Python everything works fine using mx.model.FeedForward(), in R I struggle to extract internal convolutional layers of Inception-BN in order to easily predict them on other image input. It seems I need to use Rcpp and external libraries, for example i get using your code
"missing function mx.net.slice".

**Comment by rquintino** [comment]
same issue, trying to run this transfer learning sample on R, but seems not supported yet? 
(related to allow_extra_params parameter not being available on R?)
[(http://shuokay.com/2016/09/02/transfer-learning/)](http://shuokay.com/2016/09/02/transfer-learning/)

**Comment by thirdwing** [comment]
Please try to code below. It extracted the `global_pooling_feature`, which you used later.

```r
library(mxnet)
prefix = "Inception-BN"
num.round = 126
model = mx.model.load(prefix, num.round)
internals = model$symbol$get.internals()
fea_symbol = internals[[match("global_pool_output", internals$outputs)]]

model2 <- list(symbol = fea_symbol,
              arg.params = model$arg.params,
              aux.params = model$aux.params)

class(model) <- "MXFeedForwardModel"

library(imager)

im <- load.image(system.file("extdata/parrots.png", package = "imager"))
plot(im)

preproc.image <- function(im, mean.image) {
  # crop the image
  shape <- dim(im)
  short.edge <- min(shape[1:2])
  xx <- floor((shape[1] - short.edge) / 2)
  yy <- floor((shape[2] - short.edge) / 2)
  croped <- crop.borders(im, xx, yy)
  # resize to 224 x 224, needed by input of the model.
  resized <- resize(croped, 224, 224)
  # convert to array (x, y, channel)
  arr <- as.array(resized) * 255
  dim(arr) <- c(224, 224, 3)
  # subtract the mean
  normed <- arr - mean.img
  # Reshape to format needed by mxnet (width, height, channel, num)
  dim(normed) <- c(224, 224, 3, 1)
  return(normed)
}

mean.img <- 127

normed <- preproc.image(im, mean.img)

global_pooling_feature <- predict(model, X = normed)
dim(global_pooling_feature)
## [1] 1000
```

**Comment by lichen11** [comment]
I tried your code above. 
global_pooling_feature <- predict(model, X = normed) actually gives the probability vector, as it sums up to one.
I tried using predict(model2, X = normed) and got an error. Can you shed some light on how to obtain the global pooling feature? Thanks.

**Comment by thirdwing** [comment]
I am sorry that there is a typo:

```r
library(mxnet)
prefix = "Inception-BN"
num.round = 126
model = mx.model.load(prefix, num.round)
internals = model$symbol$get.internals()
fea_symbol = internals[[match("global_pool_output", internals$outputs)]]

model2 <- list(symbol = fea_symbol,
              arg.params = model$arg.params,
              aux.params = model$aux.params)

class(model2) <- "MXFeedForwardModel"

library(imager)

im <- load.image(system.file("extdata/parrots.png", package = "imager"))
plot(im)

preproc.image <- function(im, mean.image) {
  # crop the image
  shape <- dim(im)
  short.edge <- min(shape[1:2])
  xx <- floor((shape[1] - short.edge) / 2)
  yy <- floor((shape[2] - short.edge) / 2)
  croped <- crop.borders(im, xx, yy)
  # resize to 224 x 224, needed by input of the model.
  resized <- resize(croped, 224, 224)
  # convert to array (x, y, channel)
  arr <- as.array(resized) * 255
  dim(arr) <- c(224, 224, 3)
  # subtract the mean
  normed <- arr - mean.img
  # Reshape to format needed by mxnet (width, height, channel, num)
  dim(normed) <- c(224, 224, 3, 1)
  return(normed)
}

mean.img <- 127

normed <- preproc.image(im, mean.img)

global_pooling_feature <- predict(model2, X = normed)
```

**Comment by thirdwing** [comment]
After getting the `model2`, you can set the class to be `MXFeedForwardModel` like below:

```r
model2 <- list(symbol = fea_symbol,
              arg.params = model$arg.params,
              aux.params = model$aux.params)

class(model2) <- "MXFeedForwardModel"
```

**Comment by lichen11** [comment]
I followed your instructions, but got the following error:
global_pooling_feature <- predict(model2, X = normed)
Error in exec$update.arg.arrays(arg.arrays, match.name, skip.null) : 
  executor.cc:65: RCheck failed: to->containsElementNamed(names[i].c_str()) cannot find key fc1_bias in the array arg.arrays

The image I used was one of the cat/dog images.

**Comment by thirdwing** [comment]
Can you try the code below? 

```r
library(mxnet)
prefix = "Inception-BN"
num.round = 126
model = mx.model.load(prefix, num.round)
internals = model$symbol$get.internals()
fea_symbol = internals[[match("global_pool_output", internals$outputs)]]

model$arg.params$fc1_bias <- NULL
model$arg.params$fc1_weight <- NULL

model2 <- list(symbol = fea_symbol,
              arg.params = model$arg.params,
              aux.params = model$aux.params)

class(model2) <- "MXFeedForwardModel"

library(imager)

im <- load.image(system.file("extdata/parrots.png", package = "imager"))
plot(im)

preproc.image <- function(im, mean.image) {
  # crop the image
  shape <- dim(im)
  short.edge <- min(shape[1:2])
  xx <- floor((shape[1] - short.edge) / 2)
  yy <- floor((shape[2] - short.edge) / 2)
  croped <- crop.borders(im, xx, yy)
  # resize to 224 x 224, needed by input of the model.
  resized <- resize(croped, 224, 224)
  # convert to array (x, y, channel)
  arr <- as.array(resized) * 255
  dim(arr) <- c(224, 224, 3)
  # subtract the mean
  normed <- arr - mean.img
  # Reshape to format needed by mxnet (width, height, channel, num)
  dim(normed) <- c(224, 224, 3, 1)
  return(normed)
}

mean.img <- 127

normed <- preproc.image(im, mean.img)

global_pooling_feature <- predict(model2, X = normed)

dim(global_pooling_feature)
# [1]    1    1 1024    1
```

**Comment by lichen11** [comment]
Thank you! It works. Could you explain why adding
model$arg.params$fc1_bias <- NULL
model$arg.params$fc1_weight <- NULL
would help?

**Comment by thirdwing** [comment]
`fc1_bias` and `fc1_weight` are for the last fully connected layer after pooling. (See
https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbol_inception-bn.R#L108-L115)

We don't need them if we want to get the pooling result.
